mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+        fraction_non_clifford: the fraction of non-Clifford gates to replace
+                               in the circuit of interest.
+        method_select: string defining the method used to select the
+                       non-Clifford gates to replace: 'random' or
+                       'probabilistic'.
+        method_replace: string defining method used to replace selected
+                        non-Clifford gates: 'closest', 'random' or
+                        'probabilistic'.
+        random_state: Seed for sampling.
+        sigma_select: width of probability distribution used in selection
+                      of non-Clifford gates to replace, only has effect if
+                      method_select = 'probabilistic'
+        sigma_replace: width of probability distribution used in replacement
+                       of selected non-Clifford gates, only has effect if
+                       method_replace = 'probabilistic'.
+        Returns:
+            Circuit: Near-Clifford projected circuit.
+            List[float]: list of angles replaced that were replaced in the
+                         training circuit.
+            List[float]: list of angles that were inserted in the training
+                         circuit.
+    Raises:
+        Exception: If additional options does not contain one or two keys
+                   'sigma_select' and/or 'sigma_replace' both equal to some
+                   float.
+        .
+    """
+    # set the seed for sampling, for replacement and selection:
+    (random_state_select, random_state_replace) = random_state.randint(
+        10 ** (8), size=2
+    )
+    random_state_select = np.random.RandomState(random_state_select)
+    random_state_replace = np.random.RandomState(random_state_replace)
+    # get the operations from the circuit and find the non-cliff angles:
+    operations = np.array(list(circuit.all_operations()))
+    positions = np.linspace(1, len(operations), len(operations))
+    gates = _get_gates(operations)
+    mask = np.array(
+        [isinstance(i, cirq.ops.common_gates.ZPowGate) for i in gates]
+    )
+    r_z_gates = operations[mask]
+    r_z_positions = positions[mask]
+    angles = _get_arguments(r_z_gates)
+    mask_non_cliff = ~_is_clifford_angle(angles)
+    rz_non_cliff = angles[mask_non_cliff]
+    pos_non_cliff = r_z_positions[mask_non_cliff]
+    rz_non_cliff_copy = rz_non_cliff.copy()
+    sigma_select = additional_options.setdefault("sigma_select", 0.5)
+    sigma_replace = additional_options.setdefault("sigma_replace", 0.5)
+    if (
+        "sigma_select" not in additional_options
+        and "sigma_replace" not in additional_options
+    ):
+        raise Exception(
+            "additional options must be dicitonary with \
+                        keys containing one or both of "
+            '"sigma_select" and "sigma_replace" both \
+                        equal to some positive float'
+        )
```

Comment:
Put this at the start of the function or add some spacing to improve readability.